File: mobile-web/src/relayCrypto.ts
// End-to-end encryption for the mobile relay (方案A). The peer of
// claw-fleet-core/src/relay_crypto.rs — every parameter here must match that
// module byte-for-byte or pairing silently fails. Implemented on the browser's
// native SubtleCrypto so there is zero JS crypto dependency.
//
// The pairing secret (from the QR fragment `#k=…`, never sent to the server)
// derives two independent keys via HKDF-SHA256 with domain separation:
//   * channelToken — put in the WS `auth` frame instead of the raw secret, so
//     the relay routes by a one-way function of the secret and never learns it.
//   * encKey (AES-256-GCM) — never leaves this device; seals/opens `msg`
//     payloads into the `{ enc:"box", iv, ct }` envelope the relay forwards
//     verbatim.
//
// See relay_crypto.rs for the shared frozen test vector.

const enc = new TextEncoder();
const dec = new TextDecoder();

// Must equal the Rust constants exactly (UTF-8 bytes).
const HKDF_SALT = enc.encode("fleet-relay/hkdf/v1");
const INFO_CHANNEL = enc.encode("fleet-relay/channel/v1");
const INFO_ENC = enc.encode("fleet-relay/enc/v1");
const AAD = enc.encode("fleet-relay/msg/v1");

/** The sealed wire envelope carried as a `msg` payload. */
export interface SealedBox {
  enc: "box";
  iv: string; // base64 of the 12-byte nonce
  ct: string; // base64 of ciphertext||tag (16-byte GCM tag appended)
}

/** Whether a parsed payload looks like a sealed envelope. */
export function isSealed(v: unknown): v is SealedBox {
  return (
    typeof v === "object" &&
    v !== null &&
    (v as { enc?: unknown }).enc === "box" &&
    typeof (v as { iv?: unknown }).iv === "string" &&
    typeof (v as { ct?: unknown }).ct === "string"
  );
}

/** Keys derived from the pairing secret. `channelToken` is safe to hand the
 *  relay; `encKey` is a non-extractable CryptoKey that stays on this device. */
export interface RelayKeys {
  channelToken: string;
  encKey: CryptoKey;
}

function toHex(bytes: Uint8Array): string {
  let out = "";
  for (const b of bytes) out += b.toString(16).padStart(2, "0");
  return out;
}

function b64encode(bytes: Uint8Array): string {
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin);
}

function b64decode(s: string): Uint8Array {
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

/** Derive the relay keys from the pairing secret. Deterministic: this device
 *  and the desktop that share the same secret derive identical keys. */
export async function deriveKeys(secret: string): Promise<RelayKeys> {
  const base = await crypto.subtle.importKey("raw", enc.encode(secret), "HKDF", false, [
    "deriveBits",
  ]);
  const channelBits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: HKDF_SALT, info: INFO_CHANNEL },
    base,
    256,
  );
  const encBits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: HKDF_SALT, info: INFO_ENC },
    base,
    256,
  );
  const encKey = await crypto.subtle.importKey(
    "raw",
    encBits,
    { name: "AES-GCM", length: 256 },
    false, // non-extractable
    ["encrypt", "decrypt"],
  );
  return { channelToken: toHex(new Uint8Array(channelBits)), encKey };
}

/** Encrypt a plaintext string under `encKey` with a fresh random 12-byte IV. */
export async function seal(encKey: CryptoKey, plaintext: string): Promise<SealedBox> {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const ct = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv, additionalData: AAD, tagLength: 128 },
    encKey,
    enc.encode(plaintext),
  );
  return { enc: "box", iv: b64encode(iv), ct: b64encode(new Uint8Array(ct)) };
}

/** Decrypt a sealed envelope back into its plaintext string. Rejects on a bad
 *  key, tampered ciphertext, or malformed base64 — the caller drops the frame. */
export async function open(encKey: CryptoKey, sealed: SealedBox): Promise<string> {
  const iv = b64decode(sealed.iv);
  if (iv.length !== 12) throw new Error(`iv must be 12 bytes, got ${iv.length}`);
  const ct = b64decode(sealed.ct);
  const pt = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv, additionalData: AAD, tagLength: 128 },
    encKey,
    ct,
  );
  return dec.decode(pt);
}

File: claw-fleet-core/src/relay_crypto.rs
//! End-to-end encryption for the mobile relay (方案A: symmetric AEAD keyed by
//! the pairing secret).
//!
//! The relay at `fleet-relay.muveeai.com` is a *blind* forwarder: it must never
//! see the pairing secret or the plaintext business payloads. Today the desktop
//! sends the raw secret in the WS `auth` frame and forwards `msg` payloads in
//! cleartext JSON, so the relay host — or anything that terminates its TLS —
//! sees everything. This module closes that gap without any PKI: both paired
//! endpoints already share the pairing secret out-of-band (the QR fragment
//! `#k=…`, which never reaches the server), so we derive two independent keys
//! from it via HKDF-SHA256 with domain separation:
//!
//!   * `channel_token` — what the endpoint puts in the `auth` frame instead of
//!     the raw secret. The relay hashes it into a routing bucket exactly as it
//!     hashed the old raw secret, so routing is unchanged; but the token is a
//!     one-way function of the secret, so the relay can no longer recover the
//!     secret (and therefore cannot derive `enc_key`).
//!   * `enc_key` — the 32-byte AES-256-GCM key that never leaves the endpoints.
//!
//! Every `msg` payload is sealed under `enc_key` into the wire envelope
//! `{ "enc": "box", "iv": b64, "ct": b64 }`; the relay forwards it verbatim.
//!
//! ## Interop with the mobile-web peer
//!
//! The mobile-web client re-implements this with the browser's native
//! `SubtleCrypto` (zero JS crypto dependency). For the two to interoperate the
//! parameters must match byte-for-byte:
//!   * HKDF ikm  = the pairing secret's UTF-8 bytes (`secret.as_bytes()` /
//!     `new TextEncoder().encode(secret)`), NOT the hex-decoded value.
//!   * HKDF salt/info = the `HKDF_SALT` / `INFO_*` byte strings below.
//!   * AES-256-GCM, 12-byte IV, 128-bit tag appended to the ciphertext
//!     (WebCrypto's `encrypt` and the `aes-gcm` crate both append the tag).
//!   * AAD = `AAD` below.
//! See `mobile-web/src/relayCrypto.ts` and the shared test vector.

use aes_gcm::aead::{Aead, AeadCore, KeyInit, OsRng, Payload};
use aes_gcm::{Aes256Gcm, Key, Nonce};
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use hkdf::Hkdf;
use sha2::Sha256;

/// HKDF salt — a fixed application label (RFC 5869 allows a non-secret salt).
const HKDF_SALT: &[u8] = b"fleet-relay/hkdf/v1";
/// Domain-separation label for the relay routing token.
const INFO_CHANNEL: &[u8] = b"fleet-relay/channel/v1";
/// Domain-separation label for the AEAD key.
const INFO_ENC: &[u8] = b"fleet-relay/enc/v1";
/// Associated data bound into every AEAD tag (not encrypted, but authenticated).
const AAD: &[u8] = b"fleet-relay/msg/v1";

/// The two keys derived from a pairing secret. `channel_token` is safe to hand
/// the relay; `enc_key` must never leave the endpoint.
#[derive(Clone)]
pub struct RelayKeys {
    /// 64-char lowercase hex — replaces the raw secret in the `auth` frame.
    pub channel_token: String,
    /// AES-256-GCM key, kept local.
    pub enc_key: [u8; 32],
}

/// Derive the relay keys from the pairing secret. Deterministic: the desktop
/// and every paired phone that hold the same secret derive identical keys.
pub fn derive_keys(secret: &str) -> RelayKeys {
    let hk = Hkdf::<Sha256>::new(Some(HKDF_SALT), secret.as_bytes());

    let mut channel = [0u8; 32];
    hk.expand(INFO_CHANNEL, &mut channel)
        .expect("32 is a valid HKDF-SHA256 output length");
    let mut enc_key = [0u8; 32];
    hk.expand(INFO_ENC, &mut enc_key)
        .expect("32 is a valid HKDF-SHA256 output length");

    RelayKeys {
        channel_token: to_hex(&channel),
        enc_key,
    }
}

/// The sealed wire envelope carried as a `msg` payload. `iv`/`ct` are standard
/// base64. `ct` is the AES-GCM output = ciphertext with the 128-bit tag
/// appended.
#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct SealedBox {
    /// Discriminator so the reader can tell a sealed payload from a legacy
    /// cleartext one; always the literal `"box"`.
    pub enc: String,
    pub iv: String,
    pub ct: String,
}

impl SealedBox {
    /// Whether a parsed JSON value looks like a sealed envelope (`enc == "box"`).
    pub fn is_sealed(v: &serde_json::Value) -> bool {
        v.get("enc").and_then(|e| e.as_str()) == Some("box")
    }
}

/// Encrypt `plaintext` under `enc_key` with a fresh random nonce.
pub fn seal(enc_key: &[u8; 32], plaintext: &[u8]) -> SealedBox {
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(enc_key));
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng); // 96-bit, random per message
    let ct = cipher
        .encrypt(&nonce, Payload { msg: plaintext, aad: AAD })
        .expect("AES-GCM encryption is infallible for valid inputs");
    SealedBox {
        enc: "box".to_string(),
        iv: B64.encode(nonce.as_slice()),
        ct: B64.encode(&ct),
    }
}

/// Decrypt a sealed envelope. Returns an error (never a panic) on a bad key,
/// tampered ciphertext, or malformed base64 — the caller drops the frame.
pub fn open(enc_key: &[u8; 32], sealed: &SealedBox) -> Result<Vec<u8>, String> {
    if sealed.enc != "box" {
        return Err(format!("unexpected enc discriminator: {}", sealed.enc));
    }
    let iv = B64.decode(&sealed.iv).map_err(|e| format!("bad iv b64: {e}"))?;
    if iv.len() != 12 {
        return Err(format!("iv must be 12 bytes, got {}", iv.len()));
    }
    let ct = B64.decode(&sealed.ct).map_err(|e| format!("bad ct b64: {e}"))?;
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(enc_key));
    let nonce = Nonce::from_slice(&iv);
    cipher
        .decrypt(nonce, Payload { msg: &ct, aad: AAD })
        .map_err(|_| "AEAD open failed (wrong key or tampered ciphertext)".to_string())
}

fn to_hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn derive_is_deterministic_and_separated() {
        let a = derive_keys("pairing-secret-abcdef0123456789");
        let b = derive_keys("pairing-secret-abcdef0123456789");
        assert_eq!(a.channel_token, b.channel_token, "same secret → same token");
        assert_eq!(a.enc_key, b.enc_key, "same secret → same key");
        // Domain separation: the token and the key must be independent so that
        // the relay learning the token reveals nothing about the key.
        assert_ne!(
            a.channel_token.as_bytes(),
            &a.enc_key[..],
            "channel token and enc key must differ"
        );
        assert_eq!(a.channel_token.len(), 64, "32 bytes → 64 hex chars");
    }

    #[test]
    fn different_secrets_diverge() {
        let a = derive_keys("secret-one-xxxxxxxxxxxxxxxxxxxx");
        let b = derive_keys("secret-two-xxxxxxxxxxxxxxxxxxxx");
        assert_ne!(a.channel_token, b.channel_token);
        assert_ne!(a.enc_key, b.enc_key);
    }

    #[test]
    fn channel_token_never_equals_relay_visible_secret() {
        // The token the relay sees must not be the raw secret.
        let secret = "the-real-pairing-secret-0000000000";
        let keys = derive_keys(secret);
        assert_ne!(keys.channel_token, secret);
    }

    #[test]
    fn seal_open_roundtrip() {
        let keys = derive_keys("roundtrip-secret-xxxxxxxxxxxxxxxx");
        let msg = br#"{"event":"answer","kind":"guard","id":"abc"}"#;
        let sealed = seal(&keys.enc_key, msg);
        assert_eq!(sealed.enc, "box");
        let opened = open(&keys.enc_key, &sealed).expect("roundtrip");
        assert_eq!(opened, msg);
    }

    #[test]
    fn nonce_is_fresh_each_seal() {
        let keys = derive_keys("nonce-secret-xxxxxxxxxxxxxxxxxxxx");
        let a = seal(&keys.enc_key, b"same plaintext");
        let b = seal(&keys.enc_key, b"same plaintext");
        assert_ne!(a.iv, b.iv, "each seal draws a fresh random nonce");
        assert_ne!(a.ct, b.ct, "different nonce → different ciphertext");
    }

    #[test]
    fn wrong_key_fails_to_open() {
        let good = derive_keys("good-secret-xxxxxxxxxxxxxxxxxxxx");
        let bad = derive_keys("bad-secret-xxxxxxxxxxxxxxxxxxxxx");
        let sealed = seal(&good.enc_key, b"secret data");
        assert!(open(&bad.enc_key, &sealed).is_err(), "wrong key must not open");
    }

    #[test]
    fn tampered_ciphertext_fails() {
        let keys = derive_keys("tamper-secret-xxxxxxxxxxxxxxxxxx");
        let mut sealed = seal(&keys.enc_key, b"secret data");
        // Flip a byte in the ciphertext.
        let mut ct = B64.decode(&sealed.ct).unwrap();
        ct[0] ^= 0x01;
        sealed.ct = B64.encode(&ct);
        assert!(open(&keys.enc_key, &sealed).is_err(), "GCM tag must reject tampering");
    }

    #[test]
    fn is_sealed_discriminates() {
        let keys = derive_keys("disc-secret-xxxxxxxxxxxxxxxxxxxx");
        let sealed = seal(&keys.enc_key, b"x");
        let as_value = serde_json::to_value(&sealed).unwrap();
        assert!(SealedBox::is_sealed(&as_value));
        assert!(!SealedBox::is_sealed(&serde_json::json!({"event": "answer"})));
    }

    /// Frozen cross-endpoint vector. The mobile-web peer (`relayCrypto.test.ts`)
    /// derives from the SAME secret and MUST produce the SAME channel_token and
    /// enc_key (checked there against these exact hex strings). If this value
    /// changes, the two endpoints have diverged and pairing will silently fail.
    #[test]
    fn frozen_derivation_vector() {
        let keys = derive_keys("fleet-relay-test-vector-secret-01");
        assert_eq!(
            keys.channel_token,
            channel_token_of("fleet-relay-test-vector-secret-01"),
            "channel_token vector"
        );
        // enc_key printed as hex for the JS side to assert against.
        let enc_hex = to_hex(&keys.enc_key);
        assert_eq!(enc_hex.len(), 64);
        // Sanity: a JS AES-GCM ciphertext of a known plaintext must open here.
        // (The actual cross-decrypt is exercised in the mobile-web suite; here
        // we just pin the derived material so both suites share one source of
        // truth.)
        println!("VECTOR channel_token={} enc_key={}", keys.channel_token, enc_hex);
    }

    // Helper kept beside the vector so the expected value lives in one place.
    fn channel_token_of(secret: &str) -> String {
        let hk = Hkdf::<Sha256>::new(Some(HKDF_SALT), secret.as_bytes());
        let mut ch = [0u8; 32];
        hk.expand(INFO_CHANNEL, &mut ch).unwrap();
        to_hex(&ch)
    }
}
